txpipe/utils/theory.py

```diff
@@ -51,111 +55,269 @@ def ccl_read_yaml(filename, **kwargs):
+    theory_model=default_theory_model
+):
+    """
+    Use FireCrown to generate the theory predictions for the data
+    in a sacc file.
+
+    If the supplied sacc data has only tracers and no data points then
+    the output will be given data points at the ell or theta
+    values supplued in the keyword options.
+
+    Linear galaxy bias parameters can optionally be provided as an array.
+    If left as None they are set to unity.
+
+    Parameters
+    ----------
+    cosmo: pyccl.Cosmology
+        The cosmology at which to get the theory predictions
+    sacc_data: sacc.Sacc or str
+        Sacc object or a path to one. If the file has only tracers and
+        no data then ell_values or theta_values must be supplied.
+    bias: float, array, or None
+        If a float b_0, use the scheme b = b_0 / D(<z>) for each bin.
+        If an array, use the given value for each bin
+    ell_values: array, optional
+        An array of ell's to use if no data points are in the sacc file
+    theta_values: array, optional
+        An array of theta's to use if no data points are in the sacc file
+
+    Returns
+    -------
+    sacc_theory: sacc.Sacc
+        A copy of the input sacc_data but with values
+        replaced with theory predictions.
+    """
+    from firecrown.likelihood.likelihood import load_likelihood
+    from firecrown.parameters import ParamsMap
+    import sacc
+    import pathlib
+
+    # Make sure we have computed P(k, z). If this
+    # has already been calculated then this doesn't
+    # do so again.
+    cosmo.compute_nonlin_power()
+
+    # Read the sacc data file if needed
+    if isinstance(sacc_data, (str, pathlib.Path)):
+        sacc_data = sacc.Sacc.load_fits(sacc_data)
+    else:
+        sacc_data = sacc_data.copy()
+
+    # We can optionally smooth the n(z). This helped in Prat et al.
+    # Probably this should be happening in CCL somewhere.
+    if smooth:
+        smooth_sacc_nz(sacc_data)
+
+    # The user can pass in an empty sacc file, with no data points in,
+    # if they also pass in either ell or theta values to fill it with.
+    # They can pass in either but not both.
+    if len(sacc_data.data) == 0:
+        fill_empty_sacc(sacc_data, ell_values=ell_values, theta_values=theta_values)
+
+    # Use the FireCrown machinery to compute the likelihood and as
+    # a by-product the theory
+    build_parameters = {
+        "sacc_data": sacc_data,
+    }
+
+    # These stages are a copy of what is done inside the FireCrown connectors
+    likelihood, tools = load_likelihood(theory_model, build_parameters)
+    tools.prepare(cosmo)
+
+    systematic_params = {
+        **make_bias_parameters(bias, sacc_data, cosmo)
+    }
+    # Apply the systematics parameters
+    likelihood.update(ParamsMap(systematic_params))
+
+    # Run the likelihood. As a by-product this fills in
+    # the predicted data vector
+    loglike = likelihood.compute_loglike(tools)
```

Comment:
There's now also `GaussFamily.compute_theory_vector`, which returns the same data as `predicted_data_vector` and which in the future hopefully won't require actual data or covariance in the sacc file.

As far as I can tell, the likelihood-level systematics were moved to the `ModellingTools` class, so using `compute_theory_vector` is fine.

I'm hoping to have a clean theory-only interface for firecrown at some point.

Reply:
Thanks, I didn't know about this. I've switched to using it.